relay-redis/src/real.rs

```diff
@@ -137,75 +220,116 @@ impl fmt::Debug for RedisPoolInner {
     pub fn cluster<'a>(
         servers: impl IntoIterator<Item = &'a str>,
         opts: RedisConfigOptions,
     ) -> Result<Self, RedisError> {
-        let pool = Pool::builder()
-            .max_size(opts.max_connections)
-            .min_idle(opts.min_idle)
-            .test_on_check_out(false)
-            .max_lifetime(Some(Duration::from_secs(opts.max_lifetime)))
-            .idle_timeout(Some(Duration::from_secs(opts.idle_timeout)))
-            .connection_timeout(Duration::from_secs(opts.connection_timeout))
+        let pool = Self::base_pool_builder(&opts)
             .build(redis::cluster::ClusterClient::new(servers).map_err(RedisError::Redis)?)
             .map_err(RedisError::Pool)?;
 
-        let inner = RedisPoolInner::Cluster(pool);
-        Ok(RedisPool { opts, inner })
+        Ok(RedisPool::Cluster(pool, opts))
     }
 
-    /// Creates a `RedisPool` in single-node configuration.
+    /// Creates a [`RedisPool`] in multi write configuration.
+    pub fn multi_write(
+        primary: RedisPool,
+        secondaries: Vec<RedisPool>,
+    ) -> Result<Self, RedisError> {
+        Ok(RedisPool::MultiWrite(Box::new(primary), secondaries))
+    }
+
+    /// Creates a [`RedisPool`] in single-node configuration.
     pub fn single(server: &str, opts: RedisConfigOptions) -> Result<Self, RedisError> {
-        let pool = Pool::builder()
-            .max_size(opts.max_connections)
-            .min_idle(opts.min_idle)
-            .test_on_check_out(false)
-            .max_lifetime(Some(Duration::from_secs(opts.max_lifetime)))
-            .idle_timeout(Some(Duration::from_secs(opts.idle_timeout)))
-            .connection_timeout(Duration::from_secs(opts.connection_timeout))
-            .build(redis::Client::open(server).map_err(RedisError::Redis)?)
-            .map_err(RedisError::Pool)?;
+        let pool = Self::client_pool(server, &opts)?;
 
-        let inner = RedisPoolInner::Single(pool);
-        Ok(RedisPool { opts, inner })
+        Ok(RedisPool::Single(pool, opts))
     }
 
     /// Returns a pooled connection to a client.
     pub fn client(&self) -> Result<PooledClient, RedisError> {
-        let inner = match self.inner {
-            RedisPoolInner::Cluster(ref pool) => {
-                PooledClientInner::Cluster(Box::new(pool.get().map_err(RedisError::Pool)?))
-            }
-            RedisPoolInner::Single(ref pool) => {
-                PooledClientInner::Single(Box::new(pool.get().map_err(RedisError::Pool)?))
+        let pool = match self {
+            RedisPool::Cluster(ref pool, opts) => PooledClient::Cluster(
+                Box::new(pool.get().map_err(RedisError::Pool)?),
+                opts.clone(),
+            ),
+            RedisPool::MultiWrite(ref primary_pool, ref secondary_pools) => {
+                let primary_client = primary_pool.client()?;
+                let mut secondary_clients = Vec::with_capacity(secondary_pools.len());
```

Comment:
We could use `smallvec` with a size of 1 here, we never expect more than one secondary.

Reply:
As discussed, this doesn't work unfortunately without Boxing, so let's just not do it.